---
date: 2023-11-02
---
# Можно ли просто заменить все синхронные методы асинхронными

Разработчики .NET хорошо постарались и сделали использование асинхронных методов максимально приближенным к синхронным, скрыв от программиста сложность реализации. Складывается впечатление, что можно сделать методы асинхронными и добавить в местах вызова ```await```, и все будет работать как раньше, но с преимуществом асинхронных вызовов.

На самом деле это не так и существует ряд проблем(в основном для UI приложений):

## Deadlock

Главная причина - завхват [[Synchronization Context]] при вызове асинхронного метода с целью выполнить продолжение(после выполнения асинхронной операции) в том же контексте.
Правила, которые помогут избежать блокировки:

1. Не блокировать вызывающий поток в ожидании выполенения асинхронной операции, вызывая Result или Wait().
2. В библиотечных методах использовать ```ConfigureAwait(false)``` при вызове асинхронных методов.

Можно ли вызвать ```ConfigureAwait(false)``` только на первом вызове асинхронного метода? Нет, потому что в некоторых случаях он может выполниться синхронно и тогда к следующему асинхронному методу мы подойдем все также имея ```SynchronizationContext```.

Чтобы не загромождать библиотечный код множеством ```ConfigureAwait(false)``` можно использовать ```Awaiters.DetachCurrentSyncContext()```, который делает следующее: если есть ```SynchronizationContext```, то продолжение будет выполнено в потоке, взятого из ```ThreadPool```, а если нет, то ```IsCompleted = true``` и операция будет выполнена синхронно.

## Обработка ошибок

В случае ошибок при выполнении асинхронной операции в вызывающий код пробросится:

1. ```Result```, ```Wait()``` - ```AggregateException```, содержащая все ошибки, произошедшие за время выполнения операции
1. ```async/await``` - только одна - первая. Потому что используется метод ```TaskAwaiter.GetResult()```, который также доступен ```Task.GetAwaiter().GetResult()```.

Это приводит к тому, что, во-первых, мы можем потерять часть ошибок, а, во-вторых, существующие обработчики ошибок могут быть не готовы к выбрасываемым типам.

## Асинхронные методы, возвращающие void

Когда асинхронный метод возвращает ```Task/Task<T>```, она служит как promise о том, что операций будет выполнена в будущем + по статусу задачи можно ориентироваться на статус операции. В случае ```void``` - это не получится сделать. Это приводит к тому, что в случае отсутствия ```SynchronizationContext``` при возникновении ошибки в ходе такого асинхронного метода мы получим ```TaskScheduler.UnobservedTaskException```, которое вызовет необрабатываемое ```AppDomain.UnhandledException```.

## Итог

В связи со всем вышеперечисленным вот правила:

- The continuations of async methods are scheduled into a captured synchronization context that can cause deadlocks.
- To avoid deadlocks all the async library code should be littered with ConfigureAwait(false) calls.
- await task; throws the first error making exception handling for parallel programming more complicated.
- Async void methods were introduced for handling UI events but they can be used accidentally causing the application to crash in case of an unhandled exception.

---

## Источники

1. [One user scenario to rule them all by Sergey Tepliakov](https://blogs.msdn.microsoft.com/seteplia/2018/02/14/one-user-scenario-to-rule-them-all/)

## Ссылки

1. [[Antipatterns]]
